src/event/finalization.rs

```diff
@@ -0,0 +1,336 @@
+            let finalizers = mem::replace(&mut self.0, vec![].into());
+            let mut result: Vec<_> = finalizers.into();
+            // This is the only step that may cause a (re)allocation.
+            result.reserve_exact(other.0.len());
+            // Box<[T]> is missing IntoIterator
+            let other: Vec<_> = other.0.into();
+            for entry in other.into_iter() {
+                // deduplicate by hand, assume the list is trivially small
+                if !result.iter().any(|existing| Arc::ptr_eq(existing, &entry)) {
+                    result.push(entry);
+                }
+            }
+            self.0 = result.into();
+        }
+    }
+
+    /// Update the status of all finalizers in this set.
+    pub fn update_status(&self, status: EventStatus) {
+        for finalizer in self.0.iter() {
+            finalizer.update_status(status);
+        }
+    }
+
+    /// Update all sources for this finalizer with the current
+    /// status. This *drops* the finalizer array elements so they may
+    /// immediately signal the source batch.
+    pub fn update_sources(&mut self) {
+        let finalizers = mem::replace(&mut self.0, vec![].into());
+        for finalizer in finalizers.iter() {
+            finalizer.update_batch();
+        }
+    }
+
+    #[cfg(test)]
+    fn count_finalizers(&self) -> usize {
+        self.0.len()
+    }
+}
+
+/// An event finalizer is the shared data required to handle tracking
+/// the status of an event, and updating the status of a batch with that
+/// when the event is dropped.
+#[derive(Debug)]
+pub struct EventFinalizer {
+    status: Atomic<EventStatus>,
+    batch: Arc<BatchNotifier>,
+}
+
+impl EventFinalizer {
+    /// Create a new event in a batch.
+    pub fn new(batch: Arc<BatchNotifier>) -> Self {
+        let status = Atomic::new(EventStatus::Dropped);
+        Self { status, batch }
+    }
+
+    /// Update this finalizer's status in place with the given `EventStatus`
+    pub fn update_status(&self, status: EventStatus) {
+        self.status
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |old_status| {
+                Some(old_status.update(status))
+            })
+            .unwrap_or_else(|_| unreachable!());
+    }
+
+    /// Update the batch for this event with this finalizer's
+    /// status, and mark this event as no longer requiring update.
+    pub fn update_batch(&self) {
+        let status = self
+            .status
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |_| {
+                Some(EventStatus::NoOp)
+            })
+            .unwrap_or_else(|_| unreachable!());
+        self.batch.update_status(status);
+    }
+}
+
+impl Drop for EventFinalizer {
+    fn drop(&mut self) {
+        self.update_batch();
+    }
+}
+
+/// A batch notifier contains the status of the current batch along with
+/// a one-shot notifier to send that status back to the source. It is
+/// shared among all events of a batch.
```

Comment:
It might be worth a little more exposition on the idea of a batch, since it seems to be a new concept we're introducing here. If I'm understanding correctly, these batches will be part of the API used by sources to signify sets of events that will be acknowledged as a group?

Reply:
That is exactly the case. Most sources (notably, all HTTP-based sources) receive requests that contain more than one event, and must acknowledge the entire "batch" with a single response code.